### PR TITLE
Add support for Steep (Ruby type checker)

### PR DESCRIPTION
**Pull request by ybiquitous**
[Steep](https://github.com/soutaro/steep) is a Ruby static type checker created by soutaro who is a Ruby committer and the main developer of [RBS](https://github.com/ruby/rbs).

Steep has its own language server implementation and he has created the VSCode extension for Steep.
See <https://github.com/soutaro/steep#ides>.

### CODE REVIEW

File: clients/lsp-steep.el [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+(defcustom lsp-steep-use-bundler t
+  "Run Steep using Bunder."
+  :type 'boolean
+  :safe #'booleanp
+  :group 'lsp-steep)
+
+(defun lsp-steep--build-command ()
+  "Build a command to start the Steep language server."
+  (append
```
Comment (kiennq) [inline clients/lsp-steep.el:52]: nit: you can use `nconc` here

File: clients/lsp-steep.el [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    (if lsp-steep-use-bundler '("bundle" "exec"))
+    '("steep" "langserver")
+    (list (concat "--log-level=" lsp-steep-log-level))))
+
+(lsp-register-client
+ (make-lsp-client
+  :new-connection (lsp-stdio-connection #'lsp-steep--build-command)
+  :major-modes '(ruby-mode enh-ruby-mode)
+  :priority -2
```
Comment (kiennq) [inline clients/lsp-steep.el:61]: The priority should be differnt from `sorbet-ls`

**Comment by kiennq** [comment]
Can you help consolidate `lsp-steep` with `lsp-sorbet` in the same file, for example `lsp-ruby`?
So that all language servers for `ruby` will be easily viewed in same file and don't have priority conflict.

**Comment by ybiquitous** [comment]
@kiennq I'd be happy to help! Implementing 3 tools (solargraph, sorbet, and steep) in one file sounds nice, but they have their own language server. Can we actually integrate them into one file for lsp-mode?

**Comment by kiennq** [comment]
@ybiquitous Yes, that would be fine, each of them will be in separated sections.
You can take a look at `lsp-php`, which has `php-ls`, `intelephense`, `serenata` inside it.

**Comment by ybiquitous** [comment]
Ah, I got it. I'll try it! Is it better to open a new pull request, or can I continue on this PR?

**Comment by ybiquitous** [comment]
First, it seems good to create `lsp-ruby.el` integrating `lsp-solargraph.el` and `lsp-sorbet.el` on another PR. What do you think?

**Comment by yyoncho** [comment]
@kiennq I am a bit concerned about removing lsp-solargraph.el it will be a breaking change.

**Comment by kiennq** [comment]
@yyoncho We can always put a dummy `lsp-solargraph.el` there that require `lsp-ruby` for backward compatibility and  warning when attempt to use it. WDYT?
I think we're doing the same for `lsp.el` and also removing `lsp-intelephense` when refactoring `lsp-php` too.

**Comment by yyoncho** [comment]
> @yyoncho We can always put a dummy `lsp-solargraph.el` there that require `lsp-ruby` for backward compatibility and warning when attempt to use it. WDYT?
> I think we're doing the same for `lsp.el` and also removing `lsp-intelephense` when refactoring `lsp-php` too.

In can be done (we should keep also lsp-packages references). In general, I am not big fan of grouping since the prop names became too long  `lsp-<language>-<server-id>-<property-name>`. Other than that I am fine with the change.

**Comment by kiennq** [comment]
> In general, I am not big fan of grouping since the prop names became too long `lsp-<language>-<server-id>-<property-name>`.

It  can keep the original `lsp-<server-id>-<property-name>` though.
The main reason for grouping is for easier to determine the priority and also for new contributor to know exactly where to add new language server without worrying about boiler plate code when creating Elisp file.

**Comment by ybiquitous** [comment]
@yyoncho @kiennq What status is this PR now?

**Review by yyoncho — APPROVED** [review]

**Comment by yyoncho** [comment]
Can you resolve conflicts?

File: clients/lsp-steep.el (hunk 1 above)
Comment (ybiquitous) [reply clients/lsp-steep.el:52]: The following change does not work on my environment, so I keep it as-is.

```diff
--- a/clients/lsp-steep.el
+++ b/clients/lsp-steep.el
@@ -49,7 +49,7 @@
 
 (defun lsp-steep--build-command ()
   "Build a command to start the Steep language server."
-  (append
+  (nconc
     (if lsp-steep-use-bundler '("bundle" "exec"))
     '("steep" "langserver")
     (list (concat "--log-level=" lsp-steep-log-level))))
```

File: clients/lsp-steep.el (hunk 2 above)
Comment (ybiquitous) [reply clients/lsp-steep.el]: I've changed it to `:priority -3`.

https://github.com/emacs-lsp/lsp-mode/blob/d2d346c397c507b60dd3cd4a98d79b64bcb26028/clients/lsp-sorbet.el#L52

**Comment by ybiquitous** [comment]
@yyoncho I've rebased the `master` branch and fixed the conflict. 👌

**Comment by yyoncho** [comment]
Thank you for contributing to `lsp-mode`!

**Pull request merged**